File: packages/shared-types/src/entities/space.ts
import { Persona } from "./person";

export interface EspacioPersona {
  id: string;
  espacio_id: string;
  persona_id: string;
  rol: "RESPONSABLE" | "OCUPANTE" | "BRIGADISTA" | "CONTACTO" | string;
  puesto_identificador?: string | null;
  fecha_inicio?: string | null;
  fecha_fin?: string | null;
  notas?: string | null;
  persona?: Persona | null;
  created_at: string;
  updated_at: string;
}

export interface Espacio {
  id: string;
  piso_id: string;
  codigo: string;
  nombre?: string | null;
  tipo: string;
  estado: string;
  capacidad: number;
  area_m2?: number | null;
  perimetro_m?: number | null;
  unidad_id?: string | null;
  unidad_nombre?: string | null;
  unidad_codigo?: string | null;
  sede_nombre?: string | null;
  edificio_nombre?: string | null;
  piso_numero?: number | null;
  personas_asignadas?: EspacioPersona[];
  metadata_extra?: string | null;
  created_at: string;
  updated_at: string;
}

File: apps/api/app/schemas/space.py
"""
Schemas Pydantic para Espacios Físicos y Asignaciones Espacio-Persona.
"""
import uuid
from datetime import datetime
from pydantic import BaseModel, ConfigDict
from app.schemas.person import PersonaRead


# ── EspacioPersona ────────────────────────────────────────────────────────

class EspacioPersonaBase(BaseModel):
    espacio_id: uuid.UUID
    persona_id: uuid.UUID
    rol: str = "OCUPANTE"  # RESPONSABLE, OCUPANTE, BRIGADISTA, CONTACTO
    puesto_identificador: str | None = None
    fecha_inicio: datetime | None = None
    fecha_fin: datetime | None = None
    notas: str | None = None


class EspacioPersonaCreate(EspacioPersonaBase):
    pass


class EspacioPersonaUpdate(BaseModel):
    rol: str | None = None
    puesto_identificador: str | None = None
    fecha_inicio: datetime | None = None
    fecha_fin: datetime | None = None
    notas: str | None = None


class EspacioPersonaRead(EspacioPersonaBase):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    persona: PersonaRead | None = None
    created_at: datetime
    updated_at: datetime


# ── Espacio ───────────────────────────────────────────────────────────────

class EspacioBase(BaseModel):
    piso_id: uuid.UUID
    codigo: str
    nombre: str | None = None
    tipo: str = "SALA"
    estado: str = "Disponible"
    capacidad: int = 0
    area_m2: float | None = None
    perimetro_m: float | None = None
    unidad_id: uuid.UUID | None = None
    metadata_extra: str | None = None


class EspacioCreate(EspacioBase):
    pass


class EspacioUpdate(BaseModel):
    codigo: str | None = None
    nombre: str | None = None
    tipo: str | None = None
    estado: str | None = None
    capacidad: int | None = None
    area_m2: float | None = None
    perimetro_m: float | None = None
    unidad_id: uuid.UUID | None = None
    metadata_extra: str | None = None


class EspacioRead(EspacioBase):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    created_at: datetime
    updated_at: datetime


class EspacioWithDetailsRead(EspacioRead):
    unidad_nombre: str | None = None
    unidad_codigo: str | None = None
    sede_nombre: str | None = None
    edificio_nombre: str | None = None
    piso_numero: int | None = None
    personas_asignadas: list[EspacioPersonaRead] = []
